File: mandelb.py
from typing import Tuple, List
import cProfile
from math import sqrt
from PIL import Image
import argparse as ap
from multiprocessing import Pool


Pixel = Tuple[int, int]


MAX_ITERATIONS = 1000
WIDTH = 3000
HEIGHT = 3000
THREADS = 4

# parser = ap.ArgumentParser()

# parser.add_argument('-w', '--width', type=int)


def get_mag(num: complex) -> float:
    return sqrt(num.real ** 2 + num.imag ** 2)


def scale(num, size):
    return (num - (size // 2)) / (size / 4)


def get_iterations(pixel: Pixel) -> int:
    a = scale(pixel[0], WIDTH)
    b = scale(pixel[1], HEIGHT)

    c = complex(a, b)
    z = complex(0, 0)
    iters = 0

    traversed = set()

    while get_mag(z) <= 2 and iters < MAX_ITERATIONS:
        z = z**2 + c

        if z in traversed:
            return MAX_ITERATIONS

        traversed.add(z)

        iters += 1

    return iters


def get_color(its) -> Tuple[int, int, int]:
    cols = [(0, 0, 102), (0, 0, 153), (0, 0, 204), (0, 0, 255)]

    if its == 1000:
        return (0, 0, 0)

    if its < 3:
        return cols[0]
    elif its < 7:
        return cols[1]
    elif its < 15:
        return cols[2]

    return cols[3]


def build_mandelbrot_bounds(bounds: Tuple[Pixel, Pixel]) \
        -> List[Tuple[Pixel, int]]:

    ul_bound, lr_bound = bounds

    x0, y0 = ul_bound
    x1, y1 = lr_bound

    it_list = []

    for x in range(x0, x1):
        for y in range(y0, y1):
            pixel = (x, y)
            it_list.append((pixel, get_iterations(pixel)))

    return it_list


def build_image(it_map: List[Tuple[Pixel, int]]) -> Image:
    img = Image.new('RGB', (WIDTH, HEIGHT), 'white')

    for pixel, iters in it_map:
        img.putpixel(pixel, get_color(iters))

    return img


def generateMSImage(filepath: str) -> None:
    pl = Pool(THREADS)
    bound_list = [((0, 0), (WIDTH//2, HEIGHT//2)),
                  ((WIDTH//2, 0), (WIDTH, HEIGHT//2)),
                  ((0, HEIGHT//2), (WIDTH//2, HEIGHT)),
                  ((WIDTH // 2, HEIGHT//2), (WIDTH, HEIGHT))]

    itmaps = pl.map(build_mandelbrot_bounds, bound_list)

    im = itmaps[0] + itmaps[1] + itmaps[2] + itmaps[3]

    img = build_image(im)

    img.save(filepath)


if __name__ == "__main__":
    cProfile.run(r'generateMSImage("x.png")')

File: mandelb.c
#include "mandelb.h"
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <complex.h>

int main(void)
{
    //printf("%d", get_iterations(1000, 1000, 2000, 2000, 1000));
    return 0;
}

double scale(int num, int size)
{
    //printf("%d\n", num);
    return ((num - (size / 2)) / ((double) size / 4));
}


int c_get_iterations(int x, int y, int width, int height, int max_iters)
{
    int iters = 0;

    double scaledX = scale(x, width);
    double scaledY = scale(y, height);

    double complex c = scaledX + scaledY * I;
    double complex z = 0.0 + 0.0 * I;

    
    //printf("%lf + %lf *i\n", creal(c), cimag(c));

    while ((creal(csqrt(z * z)) <= 2) && (iters < max_iters))
    { 
        z = (z * z) + c;
        //printf("%lf + %lf *i\n", creal(z), cimag(z));
        ++iters;
    }

    return iters;
}


// PixelIterations* build_mb_bounds(Pixel *ul_bound, Pixel *lr_bound)
// {
//     //printf("%d", (ul_bound -> x));

//     ssize_t size = ((lr_bound -> x) - (ul_bound -> x)) *
//     ((lr_bound -> y) - (ul_bound -> y));

//     PixelIterations **p_it_array = malloc(size * sizeof(**p_it_array));

//     int index = 0;

//     for (int x = (ul_bound -> x); x < (lr_bound -> x); ++x)
//     {
//         for (int y = (ul_bound -> y); y < (lr_bound -> y); ++y)
//         {
//             Pixel *pix = malloc(sizeof(*pix));
//             pix -> x = x;
//             pix -> y = y;

//             PixelIterations *p_it = malloc(sizeof(*p_it));
//             p_it -> pix = pix;
//             p_it -> iterations = get_iterations(pix);

//             p_it_array[index] = p_it;
//         }
//     }

//     return NULL;
// }